### ISSUE TITLE
Type parameter support on generic types

### ISSUE THREAD

**Issue by NinoFloris** [JetBrains/resharper-fsharp#347]
What was implemented in https://github.com/JetBrains/resharper-fsharp/commit/b63f18f1d1a28460021e2ab6e476405c34710522 seems to be a bit off around generic types with generic methods. (tested in EAP3, which does include the IVT fixes, thanks again for those!)

Things that are not quite working for generic types with generic methods are:
- Interface stub generation in C#, doing this still doesn't correctly produce a where clause.
- Constraint checks get shifted to the right and end up on the wrong parameter, usually causing an error on these.

Consider: 
```fs
type SomeClass<'TOuter>() =
     member _.M<'T, 'T2 when 'T :> System.Collections.Generic.IEnumerable<'T2>> () = ()
```

and called in C# like: 
```cs
    new SomeClass<string>().M<ImmutableArray<int>, int>();
```

This will produce an error on int saying that it doesn't fulfill `IEnumerable<'T2>`. The error disappears when 'TOuter is removed  from SomeClass, making it a non-generic type again.
Adding a third parameter to M (say an unconstrained 'T3), and a second param to the type (e.g. 'TOuter2) now causes the error to fall on the third parameter of M.

I suspect there to be a traversal algorithm that is off by `N = generic type parameters` somewhere which is probably the cause of both issues.

Regardless of this small issue, many thanks for the initial support! It already removed a lot of red markers from our codebase's solution tree.

**Comment by auduchinok** [comment]
> I suspect there to be a traversal algorithm that is off by N = generic type parameters somewhere which is probably the cause of both issues.

Yes, it turned out FCS returns outer type parameters as if they were defined on method.

> Regardless of this small issue, many thanks for the initial support! It already removed a lot of red markers from our codebase's solution tree.

@NinoFloris Thanks for reporting it all, it helps a lot!

**Comment by auduchinok** [comment]
Fixed in [6ad82d5](https://github.com/JetBrains/resharper-fsharp/commit/6ad82d5fd161f7be9638aa9a16bdb354272ea02b).
